crypto/core_algorithm.c

```diff
@@ -65,6 +65,7 @@ static int algorithm_do_this(OSSL_PROVIDER *provider, void *cbdata)
                 data->fn(provider, thismap, no_store, data->data);
             }
         }
+        ossl_provider_query_free(provider, map);
```

Comment:
I'm a bit surprised about the unconditional call...  maybe it's safe to do it like this, I dunno, but what I imagined was something like this:

``` diff
diff --git a/crypto/core_algorithm.c b/crypto/core_algorithm.c
index 68d6129598..8d64a7c2da 100644
--- a/crypto/core_algorithm.c
+++ b/crypto/core_algorithm.c
@@ -64,6 +64,11 @@ static int algorithm_do_this(OSSL_PROVIDER *provider, void *cbdata)
 
                 data->fn(provider, thismap, no_store, data->data);
             }
+
+            if (no_store) {
+                /* Signal to the provider that we're done with |thismap| */
+                ossl_provider_unquery_operation(provider, cur_operation);
+            }
         }
 
         /* Do we fulfill post-conditions? */
```

Reply:
What happens if no_store is false?  When do we free the array?

Reply:
We don't, then it's assumed to be constant.  Remember that `data->fn()` is allowed to remember the `OSSL_ALGORITHM` pointer if `no_cache` is false.

Reply:
I don't really see a reason to guard this with "no_store". IMO, if a provider has an "unquery" function then we should call it. If not, we don't bother.

Reply:
> We don't, then it's assumed to be constant. Remember that data->fn() is allowed to remember the OSSL_ALGORITHM pointer if no_cache is false.

Do we remember the actual OSSL_ALGORITHM, or do we take a copy of it?

Reply:
> Do we remember the actual OSSL_ALGORITHM, or do we take a copy of it?

`data->fn()` is allowed to remember the actual pointer if `no_cache` is false.  That's by design.